### PR TITLE
Add Platform InputPane interface

### LINKED ISSUE #13319: On screen keyboard
## Describe the bug
On phones the on-screen keyboard can cover up textboxes and you cannot see what you are typing. 
It is easy to add bottom margin to the scroll-viewer, BringIntoView; in GotFocus.
If there was a way of knowing how tall the On-screen-keyboard was it would be perfect, but I cannot find a solution for this problem.

## To Reproduce
have a textbox in the bottom 20% of the screen.

## Expected behavior
The textbox on top of on-screen keyboard. Or if you knew the height of the on-screen keyboard you could do this yourself.

## Environment 

IOS and Android.
Avalonia-Version: [11.0.4]

### PR DESCRIPTION
**Pull request by emmauss**
## What does the pull request do?
Adds platform api to listen for input pane, e.g. software keyboard, changes. This includes the state and bounds of the input pane.
As such, keyboard height is no longer included in android's default insets. This puts it in line with iOS implementation.
The listener can be accessed with `TopLevel.InputPane`, and if not supported by the platform, returns null.

## What is the current behavior?
On android, the software keyboard was included in the default navbar insets. This adds to the app's safe area padding, which isn't the right behavior. iOS, on the other hand, doesn't include the software keyboard in its safe area. This is expected.

## Checklist

- [ ] Added unit tests (if possible)?
- [ x ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes

This is a behavioral breaking change on android. 

## Fixed issues
Fixes #13319

### CODE REVIEW

**Comment by workgroupengineering** [comment]
What do you think about calling it `OnScreenKeyboardBound` instead of `SoftwareKeyboardRect`? Ditto for `SoftwareKeyboardState`.

Would also be useful integrate PR #13047.

**Review by maxkatz6 — APPROVED** [review]

File: samples/SafeAreaDemo/ViewModels/MainViewModel.cs [hunk 1]
```diff
@@ -1,4 +1,7 @@
-﻿using Avalonia;
+﻿using System;
+using System.Diagnostics;
```
Comment (workgroupengineering) [inline samples/SafeAreaDemo/ViewModels/MainViewModel.cs:2]:
```suggestion

```
Remove, it isn't necessary

File: src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using Avalonia.Controls;
```
Comment (workgroupengineering) [inline src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs:2]:
```suggestion

```
Remove, it isn't necessary

File: src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+using Avalonia.Controls.Platform;
+using Avalonia.MicroCom;
+using Avalonia.Win32.Interop;
+using Avalonia.Win32.Win32Com;
+
+namespace Avalonia.Win32.Input;
+
+internal unsafe class WindowsInputPane : IInputPane, IDisposable
+{
+    private static Guid CLSID_FrameworkInputPane = Guid.Parse("D5120AA3-46BA-44C5-822D-CA8092C1FC72");
```
Comment (workgroupengineering) [inline src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs:12]:
```suggestion
        // GUID: D5120AA3-46BA-44C5-822D-CA8092C1FC72
    private static readonly Guid CLSID_FrameworkInputPane = new(0xD5120AA3, 0x46BA, 0x44C5, 0x82, 0x2D, 0xCA, 0x80, 0x92, 0xC1, 0xFC, 0x72);
```

File: src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        _inputPane = UnmanagedMethods.CreateInstance<IFrameworkInputPane>(ref CLSID_FrameworkInputPane, ref SID_IFrameworkInputPane);
+
+        using (var handler = new Handler(this))
+        {
+            uint cookie = 0;
+            _inputPane.AdviseWithHWND(windowImpl.Handle.Handle, handler, &cookie);
+            _cookie = cookie;
+        }
+    }
+
```
Comment (workgroupengineering) [inline src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs:31]:
```suggestion

```

nits

File: src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    public InputPaneState State { get; private set; }
+
+    public Rect OccludedRect { get; private set; }
+
+    public event EventHandler<InputPaneStateEventArgs>? StateChanged;
+
+    private void OnStateChanged(bool showing, UnmanagedMethods.RECT? prcInputPaneScreenLocation)
+    {
+        var oldState = (OccludedRect, State);
+
```
Comment (workgroupengineering) [inline src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs:41]:
```suggestion

```
nits

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 6]
```diff
@@ -164,7 +165,7 @@ public WindowImpl()
 
             Screen = new ScreenImpl();
             _storageProvider = new Win32StorageProvider(this);
-
+            _inputPane = Win32Platform.WindowsVersion >= PlatformConstants.Windows10 ? new WindowsInputPane(this) : null;
```
Comment (workgroupengineering) [inline src/Windows/Avalonia.Win32/WindowImpl.cs:168]: there is some necessary using:
```csharp
using Avalonia.Metadata;
using Avalonia.Rendering;
using Avalonia.Win32.DirectX;
using Avalonia.Win32.WinRT;
using System.Diagnostics;
using static Avalonia.Controls.Platform.IWin32OptionsTopLevelImpl;
```
missing  _inputPane?.Dispose() in Dispose method.

**Review by workgroupengineering — APPROVED** [review]
I have try PR on Windows 10 with Capacitive touch. This works but doesn't show InputPane when TextBox got Focus. In the same machine a wpf works as expected.

It's best to useing Guid Components Constructor instead of Parse, it's about four times faster.


```

BenchmarkDotNet v0.13.10, Windows 10 (10.0.19045.3693/22H2/2022Update)
Intel Core i5-10400 CPU 2.90GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 7.0.14 (7.0.1423.51910), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.14 (7.0.1423.51910), X64 RyuJIT AVX2


```
| Method                | Mean       | Error     | StdDev    | Ratio | Allocated | Alloc Ratio |
|---------------------- |-----------:|----------:|----------:|------:|----------:|------------:|
| StringConstructor     | 37.7284 ns | 0.2268 ns | 0.1894 ns |  1.00 |         - |          NA |
| ByteArrayConstructor  |  0.6973 ns | 0.0148 ns | 0.0131 ns |  0.02 |         - |          NA |
| Parse                 | 36.7007 ns | 0.3754 ns | 0.3511 ns |  0.97 |         - |          NA |
| ComponentsConstructor |  6.1896 ns | 0.0502 ns | 0.0419 ns |  0.16 |         - |          NA |

I expected `insetsManager.IsSystemBar Visible = false` in windows to have the same behavior as setting the window to Full Screen, but it resolves to another PR.

I hope this PR will be merged into the 11.0.6 release.

**Comment by maxkatz6** [comment]
> This works but doesn't show InputPane when TextBox got Focus

It's out of the scope of this PR. Your issue is related to the IME, and not input pane events handling.

**Pull request merged**